cpp/src/arrow/python/numpy_convert.cc

```diff
@@ -274,5 +275,333 @@ Status TensorToNdarray(const std::shared_ptr<Tensor>& tensor, PyObject* base,
+
+  std::shared_ptr<Buffer> data = std::make_shared<NumPyBuffer>(ao);
+  std::vector<int64_t> shape(ndim);
+  std::vector<int64_t> strides(ndim);
+
+  {
+    PyAcquireGIL lock;
+    npy_intp* array_strides = PyArray_STRIDES(ndarray);
+    npy_intp* array_shape = PyArray_SHAPE(ndarray);
+    for (int i = 0; i < ndim; ++i) {
+      if (array_strides[i] < 0) {
+        return Status::Invalid("Negative ndarray strides not supported");
+      }
+      shape[i] = array_shape[i];
+      strides[i] = array_strides[i];
+    }
+
+    std::shared_ptr<DataType> type;
+    RETURN_NOT_OK(
+        GetTensorType(reinterpret_cast<PyObject*>(PyArray_DESCR(ndarray)), &type));
+    Tensor tensor(type, data, shape, strides, dim_names);
+    *out = std::make_shared<SparseTensorCOO>(tensor);
+    return Status::OK();
+  }
+}
+
+Status DenseNdarrayToSparseTensorCSR(MemoryPool* pool, PyObject* ao,
+                                     const std::vector<std::string>& dim_names,
+                                     std::shared_ptr<SparseTensorCSR>* out) {
+  if (!PyArray_Check(ao)) {
+    return Status::TypeError("Did not pass ndarray object");
+  }
+
+  PyArrayObject* ndarray = reinterpret_cast<PyArrayObject*>(ao);
+
+  int ndim = PyArray_NDIM(ndarray);
+
+  std::shared_ptr<Buffer> data = std::make_shared<NumPyBuffer>(ao);
+  std::vector<int64_t> shape(ndim);
+  std::vector<int64_t> strides(ndim);
+
+  {
+    PyAcquireGIL lock;
+    npy_intp* array_strides = PyArray_STRIDES(ndarray);
+    npy_intp* array_shape = PyArray_SHAPE(ndarray);
+    for (int i = 0; i < ndim; ++i) {
+      if (array_strides[i] < 0) {
+        return Status::Invalid("Negative ndarray strides not supported");
+      }
+      shape[i] = array_shape[i];
+      strides[i] = array_strides[i];
+    }
+
+    std::shared_ptr<DataType> type;
+    RETURN_NOT_OK(
+        GetTensorType(reinterpret_cast<PyObject*>(PyArray_DESCR(ndarray)), &type));
+    Tensor tensor(type, data, shape, strides, dim_names);
+    *out = std::make_shared<SparseTensorCSR>(tensor);
+    return Status::OK();
+  }
+}
+
+Status NdarraysToSparseTensorCOO(MemoryPool* pool, PyObject* data_ao, PyObject* coords_ao,
+                                 const std::vector<int64_t>& shape,
+                                 const std::vector<std::string>& dim_names,
+                                 std::shared_ptr<SparseTensorCOO>* out) {
+  if (!PyArray_Check(data_ao) || !PyArray_Check(coords_ao)) {
+    return Status::TypeError("Did not pass ndarray object");
+  }
+
+  PyAcquireGIL lock;
+
+  PyArrayObject* ndarray_data = reinterpret_cast<PyArrayObject*>(data_ao);
+  PyArrayObject* ndarray_coords = reinterpret_cast<PyArrayObject*>(coords_ao);
+
+  std::shared_ptr<Buffer> data = std::make_shared<NumPyBuffer>(data_ao);
+  std::shared_ptr<Buffer> coords_buffer = std::make_shared<NumPyBuffer>(coords_ao);
+
+  int coords_ndim = PyArray_NDIM(ndarray_coords);
+
+  std::shared_ptr<DataType> type;
+  const int64_t i64_size = sizeof(int64_t);
```

Comment:
You're assuming the Numpy arrays has a `int64` dtype, but you're not checking for it...

Reply:
Oh, yeah, got it.
I'm now [checking type in c++](https://github.com/apache/arrow/blob/5c66ceee02c18853eea4a55e991b4c086b21f144/cpp/src/arrow/python/numpy_convert.cc#L515) and [converting to `int64` in Cython](https://github.com/apache/arrow/blob/5c66ceee02c18853eea4a55e991b4c086b21f144/python/pyarrow/array.pxi#L925) if user passes the wrong type.